golubevm/lab4/2/mmap-function.c

```diff
@@ -0,0 +1,69 @@
+#include <stdio.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#define PAGE_SIZE 4096
+#define PAGES 10
+#define SLEEP 15
+
+
+void get_pid();
+char *get_mapped_area();
+void change_mapped_memory_protection_to_none(char *m_begin);
+void try_to_read_from_protected_memory(char *mem);
+void try_to_write_to_protected_memory(char *mem);
+void sigsegv_handler(int signo);
+
+int main(int argc, char const *argv[])
+{
+	get_pid();
+
+	char *mapped_area = get_mapped_area();
+	// sleep(SLEEP);
+
+	change_mapped_memory_protection_to_none(mapped_area);
+	// sleep(SLEEP);
+
+	signal(SIGSEGV, sigsegv_handler);
+	// try_to_read_from_protected_memory(mapped_area + 10);
+	try_to_write_to_protected_memory(mapped_area + 5);
+
+	return 0;
+}
+
+void get_pid() {
+	printf("pid: %d\n", getpid());
+	// sleep(SLEEP);
+}
+
+char *get_mapped_area() {
+	char *ptr = mmap(NULL, PAGE_SIZE * PAGES, PROT_EXEC|PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+	printf("mapped area: %p\n", ptr);
+	return ptr;
+}
+
+void change_mapped_memory_protection_to_none(char *m_begin) {
+	int err = mprotect(m_begin, PAGE_SIZE, PROT_NONE);
+	if (err != 0) {
+		perror(strerror(errno));
+		exit(1);
+	}
+}
+
+void try_to_read_from_protected_memory(char *mem) {
+	char rd = *mem;
+}
+
+void try_to_write_to_protected_memory(char *mem) {
+	*mem = 'a';
+}
+
+void sigsegv_handler(int signo) {
+	printf("Segmenatation fault occured.\n");

```

Comment:
Нужно использовать здесь `write`:
> When performing buffered I/O on a file, the stdio functions must maintain a statically allocated data buffer along with associated counters and indexes (or pointers) that record the amount of data and the current position in the buffer.  Suppose that the main program is in the middle of a call to a stdio function such as [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html) where the buffer and associated variables have been partially updated.  If, at that moment, the program is interrupted by a signal handler that also calls [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html), then the second call to [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html) will operate on inconsistent data, with unpredictable results.

Или
> The primary problem is that if the signal interrupts malloc() or some similar function, the internal state may be temporarily inconsistent while it is moving blocks of memory between the free and used list, or other similar operations. If the code in the signal handler calls a function that then invokes malloc(), this may completely wreck the memory management.